lib/frame.c

```diff
@@ -65,14 +65,16 @@ uint8_t *quicly_encode_ack_frame(uint8_t *dst, uint8_t *dst_end, quicly_ranges_t
 
 int quicly_decode_ack_frame(const uint8_t **src, const uint8_t *end, quicly_ack_frame_t *frame, int is_ack_ecn)
 {
-    uint64_t i, tmp;
+    uint64_t i, tmp, curr_gap;
 
     if ((frame->largest_acknowledged = quicly_decodev(src, end)) == UINT64_MAX)
         goto Error;
     if ((frame->ack_delay = quicly_decodev(src, end)) == UINT64_MAX)
         goto Error;
     if ((frame->num_gaps = quicly_decodev(src, end)) == UINT64_MAX)
         goto Error;
+    if (frame->num_gaps > QUICLY_ACK_MAX_GAPS)
+        frame->num_gaps = QUICLY_ACK_MAX_GAPS;
```

Comment:
I am not sure if this change is correct.

If I understand correctly, this change would cause the decode function stop after reading certain number of gaps, causing the decode of next frame to fail.

Instead, what we should be doing is skip the excessive ack ranges, pretending as if they did not exist.

Reply:
I think I see what you mean. This new line contradicts the PR description, which is a sad mistake.

> We should decode the entire ACK Frame, but only store the ACK Ranges that we can

Is my understanding correct that `num_gaps` should instead be set after the decode loop is complete?

Furthermore I think this exercise has made me realize that my change in the decoder itself is inadequate: `smallest_acknowledged` is currently freely modified with any observed range in the frame. I believe `smallest_acknowledged` should only be modified by ranges that we can actually store, otherwise the final value is misleading/incorrect in respect to the gap/range array content.